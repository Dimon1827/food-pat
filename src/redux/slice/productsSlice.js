import { createSlice } from '@reduxjs/toolkit';

const initialState = [];

const productsSlice = createSlice({
  name: 'products',
  initialState,
  reducers: {
    addProduct: (state, action) => {
      state.push(action.payload);
    },
    deleteProduct: (state, action) => {
      return state.filter(({ id }) => action.payload !== id);
    },
    clearAll: () => {
      return [];
    },
  },
});

export const { addProduct, deleteProduct, clearAll } = productsSlice.actions;

export const selectProducts = (state) => state.products;

export default productsSlice.reducer;
