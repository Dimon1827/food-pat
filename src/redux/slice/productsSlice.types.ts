import { IProduct } from '@/types';

export interface IProducts
  extends Omit<IProduct, 'ingredients' | 'description'> {
  isChecked: boolean;
  quantity: number;
}

export interface IReduxProduct {
  products: IProducts[];
  totalPrice: number;
  totalQuantity: number;
}
