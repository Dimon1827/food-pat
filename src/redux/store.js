import { configureStore, combineReducers } from '@reduxjs/toolkit';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import productsReducer from './slice/productsSlice';

const persistConfig = {
  key: 'root',
  version: 2,
  storage,
};

const reducer = combineReducers({
  products: productsReducer,
});

const persistedReducer = persistReducer(persistConfig, reducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
});

export const persistor = persistStore(store);

export default store;
