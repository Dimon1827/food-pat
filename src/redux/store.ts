import { configureStore, combineReducers } from '@reduxjs/toolkit';
import { useDispatch } from 'react-redux';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import productsReducer from './slice/productsSlice';

const persistConfig = {
  key: 'root',
  version: 2,
  storage,
};

const reducer = combineReducers({
  products: productsReducer,
});

const persistedReducer = persistReducer(persistConfig, reducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
});

export type RootState = ReturnType<typeof persistedReducer>;
export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const persistor = persistStore(store);
export default store;
