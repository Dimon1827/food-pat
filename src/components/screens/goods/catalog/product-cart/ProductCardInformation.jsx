import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { addProduct } from '../../../../../redux/slice/productsSlice';

function ProductCardInformation({
  name,
  price,
  ingredients,
  oldPrice,
  id,
  image,
}) {
  const dispatch = useDispatch();

  const handlerAddProductClick = () => {
    const product = { id, name, price, image, isChecked: false };
    dispatch(addProduct(product));
  };
  return (
    <div className="product-card__wrapper">
      <div className="product-card__top-wrapper">
        <Link className="product-card__link" to={name}>
          <h2 className="product-card__subtitle subtitle">{name}</h2>
        </Link>
        <button
          className="product-card__button button"
          onClick={handlerAddProductClick}
        >
          Купить
        </button>
      </div>
      <p className="product-card__text">Ингредиенты:</p>
      <p className="product-card__ingredients">{ingredients}</p>
      <p className="product-card__text">Цена:</p>
      {!!oldPrice && (
        <span className="food__definition additional-products__old-price">
          {oldPrice} ₽
        </span>
      )}
      <span className="food__definition food__price">{price} ₽</span>
    </div>
  );
}

export default ProductCardInformation;
