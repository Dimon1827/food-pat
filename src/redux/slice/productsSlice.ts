import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IProducts, IReduxProduct } from './productsSlice.types';

const initialState: IReduxProduct = {
  products: [],
  totalPrice: 0,
  totalQuantity: 0,
};

const productsSlice = createSlice({
  name: 'products',
  initialState,
  reducers: {
    addProduct: (state, action: PayloadAction<IProducts>) => {
      state.products.push(action.payload);
    },
    deleteProduct: (state, action: PayloadAction<string>) => {
      return {
        ...state,
        products: state.products.filter(({ id }) => action.payload !== id),
      };
    },
    toggleProduct: (state, action: PayloadAction<string>) => {
      return {
        ...state,
        products: state.products.map((product) => {
          return product.id === action.payload
            ? { ...product, isChecked: !product.isChecked }
            : product;
        }),
      };
    },
    deleteSelectedProducts: (state) => {
      return {
        ...state,
        products: state.products.filter((product) => !product.isChecked),
      };
    },
    clearAll: () => {
      return {
        products: [],
        totalPrice: 0,
        totalQuantity: 0,
      };
    },
    incrementQuantity: (state, action: PayloadAction<string>) => {
      const product = state.products.find((item) => item.id === action.payload);
      if (product) {
        product.quantity++;
      }
    },
    decrementQuantity: (state, action: PayloadAction<string>) => {
      const product = state.products.find((item) => item.id === action.payload);
      if (product && product.quantity > 1) {
        product.quantity--;
      }
    },
    calculateAll: (state) => {
      const { total, quantity } = state.products.reduce(
        (productsTotal, product) => {
          const { price, quantity } = product;
          const productTotal = price * quantity;

          productsTotal.total += productTotal;
          productsTotal.quantity += quantity;

          return productsTotal;
        },
        { total: 0, quantity: 0 }
      );
      state.totalPrice = total;
      state.totalQuantity = quantity;
    },
  },
});

export const {
  addProduct,
  deleteProduct,
  clearAll,
  toggleProduct,
  deleteSelectedProducts,
  incrementQuantity,
  decrementQuantity,
  calculateAll,
} = productsSlice.actions;

export default productsSlice.reducer;
