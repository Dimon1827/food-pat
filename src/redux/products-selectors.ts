import { RootState } from "./store";

export const selectProducts = (state: RootState) => state.products.products;
export const selectTotalPrice = (state: RootState) => state.products.totalPrice;
export const selectTotalQuantity = (state: RootState) => state.products.totalQuantity;
