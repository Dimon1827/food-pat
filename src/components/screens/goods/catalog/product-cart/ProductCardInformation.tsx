import { Link } from 'react-router-dom';
import { addProduct, calculateAll } from '../../../../../redux/slice/productsSlice';
import { IProduct } from 'types';
import { useAppDispatch } from '../../../../../redux/store';
import productWithQuantity from '../../../../../utils/productWithQuantity';



interface IProductCardInformation extends Omit<IProduct, 'description'> {}

function ProductCardInformation({
  name,
  price,
  ingredients,
  oldPrice,
  id,
  image,
}: IProductCardInformation) {
  const dispatch = useAppDispatch();

  const handleAddProductClick = () => {
    const product = { id, name, price, image, oldPrice };
    dispatch(addProduct(productWithQuantity(product)));
    dispatch(calculateAll());
  };
  return (
    <div className="product-card__wrapper">
      <div className="product-card__top-wrapper">
        <Link className="product-card__link" to={name}>
          <h2 className="product-card__subtitle subtitle">{name}</h2>
        </Link>
        <button
          className="product-card__button button"
          onClick={handleAddProductClick}
        >
          Купить
        </button>
      </div>
      <p className="product-card__text">Ингредиенты:</p>
      <p className="product-card__ingredients">{ingredients}</p>
      <p className="product-card__text">Цена:</p>
      {!!oldPrice && (
        <span className="food__definition additional-products__old-price">
          {oldPrice} ₽
        </span>
      )}
      <span className="food__definition food__price">{price} ₽</span>
    </div>
  );
}

export default ProductCardInformation;
