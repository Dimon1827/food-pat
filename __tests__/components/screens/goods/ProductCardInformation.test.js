import ProductCardInformation from '@/components/screens/goods/catalog/product-cart/ProductCardInformation';
import * as actions from '@/redux/slice/productsSlice';
import * as store from '@/redux/store';
import { renderWithRouter } from '@/utils/renderWithRouter';
import { fireEvent, screen } from '@testing-library/react';

const mockedDispatch = jest.spyOn(store, 'useAppDispatch');

const basketProductInformation = {
  id: '123',
  name: 'Сок из апельсина',
  price: '120',
  oldPrice: '200',
  image: './assets/images/day-item-2',
  isChecked: false,
  quantity: 1,
};

const productInformation = {
  name: 'Сок из апельсина',
  ingredients: 'Апельсин, грейфрукт',
  price: '120',
  oldPrice: '200',
  image: './assets/images/day-item-2',
};

const productWithoutOldPriceInformation = {
  name: 'Сок из апельсина',
  ingredients: 'Апельсин, грейфрукт',
  price: '120',
  image: './assets/images/day-item-2',
};

describe('ProductCardInformation', () => {
  it('ProductCardInformation renders correctly', () => {
    mockedDispatch.mockReturnValue(jest.fn());
    renderWithRouter(<ProductCardInformation {...productInformation} />);
    expect(screen.getByTestId('product-card-information')).toMatchSnapshot();
  });
  it('should dispatch actions', () => {
    const dispatch = jest.fn();
    mockedDispatch.mockReturnValue(dispatch);
    const mockedAddProduct = jest.spyOn(actions, 'addProduct');
    const mockedCalculateAll = jest.spyOn(actions, 'calculateAll');

    renderWithRouter(
      <ProductCardInformation id="123" {...productInformation} />
    );
    fireEvent.click(screen.getByRole('button'));
    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(mockedAddProduct).toHaveBeenCalledWith(basketProductInformation);
    expect(mockedCalculateAll).toHaveBeenCalledTimes(1);
  });
  it('ProductCardInformation doesnt have information about old price', () => {
    mockedDispatch.mockReturnValue(jest.fn());
    renderWithRouter(
      <ProductCardInformation {...productWithoutOldPriceInformation} />
    );
    expect(screen.queryByText('200 ₽')).not.toBeInTheDocument();
  });
});
