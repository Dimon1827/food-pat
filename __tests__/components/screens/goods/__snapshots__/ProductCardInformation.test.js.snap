// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCardInformation ProductCardInformation renders correctly 1`] = `
<div
  class="product-card__wrapper"
  data-testid="product-card-information"
>
  <div
    class="product-card__top-wrapper"
  >
    <a
      class="product-card__link"
      href="/Сок из апельсина"
    >
      <h2
        class="product-card__subtitle subtitle"
      >
        Сок из апельсина
      </h2>
    </a>
    <button
      class="product-card__button button"
    >
      Купить
    </button>
  </div>
  <p
    class="product-card__text"
  >
    Ингредиенты:
  </p>
  <p
    class="product-card__ingredients"
  >
    Апельсин, грейфрукт
  </p>
  <p
    class="product-card__text"
  >
    Цена:
  </p>
  <span
    class="food__definition additional-products__old-price"
  >
    200
     ₽
  </span>
  <span
    class="food__definition food__price"
  >
    120
     ₽
  </span>
</div>
`;
